common/log.man
.\" @(#)$RCSfile: log.man,v $ $Revision: 1.5 $ $Date: 2001/09/26 10:03:15 $ CERN IT-PDP/DC Frederic Hemmer
.\" Copyright (C) 1990-2000 by CERN/IT/PDP/DC
.\" All rights reserved
.\"
.TH LOG 3 "$Date: 2001/09/26 10:03:15 $" CASTOR "Common Library Functions"
.SH NAME
log, initlog  \- log messages
.SH SYNOPSIS
.nf
.B #include <log.h>
.LP
.B initlog(ident, level, output)
.B char \(**ident;
.B int \(*level;
.B char \(**output;
.LP
.B log(level, format, message, parameters, \&.\|.\|. )
.B int \(*level;
.B char \(**format;
.B char \(**message;
.fi
.IX  "openlog function"  ""  "\fLinitlog\fP \(em initialize log file"
.IX  "log function"  ""  "\fLlog\fP \(em write message to log file"
.SH DESCRIPTION
.LP
.B log(\|)
logs
.I message
to
an appropriate log file, opened by
.BR initlog (l).
on another host over the network.
The message is tagged with a priority of
.IR level .
The message looks like a
.BR printf (3S)
string.
Priorities are encoded as a
.IR level .
The level is selected from an ordered list:
.RS
.TP 20
.SB LOG_EMERG
A panic condition.  This is normally broadcast to all users.
.TP
.SB LOG_ALERT
A condition that should be corrected immediately,
such as a corrupted system database.
.TP
.SB LOG_CRIT
Critical conditions, such as hard device errors.
.TP
.SB LOG_ERR
Errors.
.TP
.SB LOG_WARNING
Warning messages.
.TP
.SB LOG_NOTICE
Conditions that are not error conditions,
but that may require special handling.
.TP
.SB LOG_INFO
Informational messages.
.TP
.SB LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.RE
.LP
If special processing is needed,
.B initlog(\|)
can be called to initialize the log file.
The parameter
.I ident
is a string that is prepended to every message.
.I level
is a indicating the level logging.
.RE
.SH COMMENT
The total line length printed should not exceed \fBBUFSIZ\fP-1 characters. When you provide a format string and arguments to
.B log(\|)
function, you have to be aware that 
.B log(\|)
function prepend each line with a string that have the format
.ft CW
.nf
.sp
"%s %s[%d]: "
.ft
.LP
.fi
where first string is a timestamp with the format
.ft CW
.nf
.sp
"%b %d %H:%M:%S" or "%b %e %H:%M:%S"
.ft
.LP
.fi
 on Windows or Unix, respectively, the second string is a generic logname, defined previously using
.B initlog(\|)
function, then the decimal number is the process id of the caller.
.br
In case of multi-threaded programs, the string will be prepended with "%s %s[%d,%d]: ", where the second number is an internal thread id specific to CASTOR applications, or -1 if not found.
.P
This is why you should know in advance the order of magnitude of the length of your message.
.SH EXAMPLE
In case of non-multithreaded program you would see:
.ft CW
.nf
.sp
Nov  1 17:34:27 progname[1234]: This is your message with possible arguments, your pid is 1234
.ft
.LP
.fi
.P
In case of a multi-threaded application you would see:
.ft CW
.nf
.sp
Sep 26 11:40:24 progname[1234,4]: This is another message with other possible arguments, your pid is 1234, CASTOR's Thread Id is 4
.ft
.LP
.fi
.SH "SEE ALSO"
syslog(3), Cthread(3), stdio(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


castor/log/Log.cpp
/******************************************************************************
 *                      Log.cpp
 *
 * This file is part of the Castor project.
 * See http://castor.web.cern.ch/castor
 *
 * Copyright (C) 2003  CERN
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * Interface to the CASTOR logging system
 *
 * @author Sebastien Ponce
 *****************************************************************************/

// Include Files
#include "castor/log/Log.hpp"
#include "h/Castor_limits.h"
#include "h/getconfent.h"

#include <errno.h>
#include <pthread.h>
#include <string.h>
#include <sys/un.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/syscall.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>

namespace {

/**
 * Default size of a syslog message.
 */
const size_t DEFAULT_SYSLOG_MSGLEN = 1024;

/**
 * Default size of a rsyslog message.
 */
const size_t DEFAULT_RSYSLOG_MSGLEN = 2000;

/**
 * Maximum length of a parameter name.
 */
const size_t LOG_MAX_PARAMNAMELEN = 20;

/**
 * Maximum length of a string value.
 */
const size_t LOG_MAX_PARAMSTRLEN = 1024;

/**
 * Maximum length of the program name to be prepended to every log message.
 */
const size_t LOG_MAX_PROGNAMELEN = 20;

/**
 * Maximum length of a log message.
 */
const size_t LOG_MAX_LINELEN = 8192;

/**
 * True if the interface of the CASTOR logging system has been initialized.
 */
bool s_initialized = false;

/**
 * The file descriptor of the log.
 */
int s_logFile = -1;

/**
 * AF_UNIX address of local logger.
 */
struct sockaddr_un s_syslogAddr;

/**
 * True if the the syslog connect has bveen done.
 */
bool s_connected = false;

/**
 * Mutex used to synchronize syslog client.
 */
pthread_mutex_t s_syslogMutex;

/**
 * The name of the program to be prepended to every log message.
 */
const char* s_progname = 0;

/**
 * The maximum message size that the client syslog server can handle.
 */
int s_maxmsglen = DEFAULT_SYSLOG_MSGLEN;

struct PrioritySyslogMapping {
  const char *name; // Name of the priority
  int  value;       // The priority's numeric representation in syslog
  const char *text; // Textual representation of the priority
};

PrioritySyslogMapping s_prioritylist[] = {
  { "LOG_EMERG",   LOG_EMERG,   "Emerg"  },
  { "LOG_ALERT",   LOG_ALERT,   "Alert"  },
  { "LOG_CRIT",    LOG_CRIT,    "Crit"   },
  { "LOG_ERR",     LOG_ERR,     "Error"  },
  { "LOG_WARNING", LOG_WARNING, "Warn"   },
  { "LOG_NOTICE",  LOG_NOTICE,  "Notice" },
  { "LOG_INFO",    LOG_INFO,    "Info"   },
  { "LOG_DEBUG",   LOG_DEBUG,   "Debug"  },
  { NULL,          0,           NULL     }
};

//-----------------------------------------------------------------------------
// castor_openlog
//-----------------------------------------------------------------------------
void castor_openlog() {
  if(-1 == s_logFile) {
    s_syslogAddr.sun_family = AF_UNIX;
    strncpy(s_syslogAddr.sun_path, _PATH_LOG, sizeof(s_syslogAddr.sun_path));
    s_logFile = socket(AF_UNIX, SOCK_DGRAM, 0);
    fcntl(s_logFile, F_SETFD, FD_CLOEXEC);
    if(-1 == s_logFile) {
      return;
    }
  }
  if(!s_connected) {
    if(-1 == connect(s_logFile, (struct sockaddr *)&s_syslogAddr,
      sizeof(s_syslogAddr))) {
      close(s_logFile);
      s_logFile = -1;
    } else {
#ifdef __APPLE__
      // MAC has has no MSG_NOSIGNAL
      // but >= 10.2 comes with SO_NOSIGPIPE
      int set = 1;
      if (0 != setsockopt(s_logFile, SOL_SOCKET, SO_NOSIGPIPE, &set,
        sizeof(int))) {
        close(s_logFile);
        s_logFile = -1;
        return;
      }
#endif
      s_connected = true;
    }
  }
}

//-----------------------------------------------------------------------------
// castor_closelog
//-----------------------------------------------------------------------------
void castor_closelog() {
  if (!s_connected) return;
  close (s_logFile);
  s_logFile = -1;
  s_connected = false;
}

//-----------------------------------------------------------------------------
// castor_syslog
//-----------------------------------------------------------------------------
void castor_syslog(const char *const msg, const int msgLen) throw() {
  int send_flags = 0;
#ifndef __APPLE__
  // MAC has has no MSG_NOSIGNAL
  // but >= 10.2 comes with SO_NOSIGPIPE
  send_flags = MSG_NOSIGNAL;
#endif
  // enter critical section
  pthread_mutex_lock(&s_syslogMutex);

  // Get connected, output the message to the local logger.
  if (!s_connected) {
    castor_openlog();
  }

  if (!s_connected ||
    send(s_logFile, msg, msgLen, send_flags) < 0) {
    if (s_connected) {
      // Try to reopen the syslog connection.  Maybe it went down.
      castor_closelog();
      castor_openlog();
    }
    if (!s_connected ||
      send(s_logFile, msg, msgLen, send_flags) < 0) {
      castor_closelog();  // attempt re-open next time
    }
  }

  // End of critical section.
  pthread_mutex_unlock(&s_syslogMutex);
}

//-----------------------------------------------------------------------------
// build_syslog_header
//-----------------------------------------------------------------------------
int build_syslog_header(char *const buffer,
                               const int buflen,
                               const int priority,
                               const struct timeval &timeStamp,
                               const int pid) {
  struct tm tmp;
  int len = snprintf(buffer, buflen, "<%d>", priority);
  localtime_r(&(timeStamp.tv_sec), &tmp);
  len += strftime(buffer + len, buflen - len, "%Y-%m-%dT%T", &tmp);
  len += snprintf(buffer + len, buflen - len, ".%06ld",
    (unsigned long)timeStamp.tv_usec);
  len += strftime(buffer + len, buflen - len, "%z: ", &tmp);
  // dirty trick to have the proper timezone format (':' between hh and mm)
  buffer[len-2] = buffer[len-3];
  buffer[len-3] = buffer[len-4];
  buffer[len-4] = ':';
  // if no source given, you by default the name of the process in which we run
  // print source and pid
  len += snprintf(buffer + len, buflen - len, "%s[%d]: ", s_progname, pid);
  return len;
}

//-----------------------------------------------------------------------------
// _clean_string
//-----------------------------------------------------------------------------
std::string _clean_string(const std::string &s, const bool underscore) {
  char *str = strdup(s.c_str());

  // Return an empty string if the strdup() failed
  if(NULL == str) {
    return "";
  }

  // Variables
  char *end = NULL;
  char *ptr = NULL;

  // Replace newline and tab with a space
  while (((ptr = strchr(str, '\n')) != NULL) ||
         ((ptr = strchr(str, '\t')) != NULL)) {
    *ptr = ' ';
  }

  // Remove leading whitespace
  while (isspace(*str)) str++;

  // Remove trailing whitespace
  end = str + strlen(str) - 1;
  while (end > str && isspace(*end)) end--;

  // Write new null terminator
  *(end + 1) = '\0';

  // Replace double quotes with single quotes
  while ((ptr = strchr(str, '"')) != NULL) {
    *ptr = '\'';
  }

  // Check for replacement of spaces with underscores
  if (underscore) {
    while ((ptr = strchr(str, ' ')) != NULL) {
      *ptr = '_';
    }
  }

  std::string result(str);
  free(str);
  return result;
}

} // unnamed namespace

//-----------------------------------------------------------------------------
// initLog
//-----------------------------------------------------------------------------
int castor::log::initLog(const char *const progname) {

  /* Variables */
  FILE *fp = NULL;
  char *p;
  char buffer[1024];
  size_t size = 0;

  /* Check if already initialized */
  if (s_initialized) {
    errno = EPERM;
    return (-1);
  }

  /* Check if the progname is too big */
  if (strlen(progname) > LOG_MAX_PROGNAMELEN) {
    errno = EINVAL;
    return (-1);
  }

  s_progname = progname;

  /* Determine the maximum message size that the client syslog server can
   * handle.
   */
  if ((p = getconfent("LOG", "MaxMessageSize", 0)) != NULL) {
    size = atoi(p);
  } else {
    /* Determine the size automatically, this is not guaranteed to work! */
    fp = fopen("/etc/rsyslog.conf", "r");
    if (fp) {
      /* The /etc/rsyslog.conf file exists so we assume the default message
       * size of 2K.
       */
      s_maxmsglen = DEFAULT_RSYSLOG_MSGLEN;

      /* In rsyslog versions >= 3.21.4, the maximum size of a message became
       * configurable through the $MaxMessageSize global config directive.
       * Here we attempt to find out if the user has increased the size!
       */
      while (fgets(buffer, sizeof(buffer), fp) != NULL) {
        if (strncasecmp(buffer, "$MaxMessageSize", 15)) {
          continue; /* Option not of interest */
        }
        size = atol(&buffer[15]);
      }
      fclose(fp);
    } else {
      /* The /etc/rsyslog.conf file is missing which implies that we are
       * running on a stock syslogd system, therefore the message size is
       * governed by the syslog RFC: http://www.faqs.org/rfcs/rfc3164.html
       */
    }
  }

  /* Check that the size of messages falls within acceptable limits */
  if ((size >= DEFAULT_SYSLOG_MSGLEN) && (size <= LOG_MAX_LINELEN)) {
    s_maxmsglen = size;
  }

  /* create the syslog serialization lock */
  if (pthread_mutex_init(&s_syslogMutex, NULL)) {
    errno = ENOMEM;
    return (-1);
  }

  /* Open syslog */
  castor_openlog();
  s_initialized = 1;

  return (0);
}

//-----------------------------------------------------------------------------
// writeMsg
//-----------------------------------------------------------------------------
void castor::log::writeMsg(
  const int priority,
  const std::string &msg,
  const int numParams,
  const castor::log::Param params[],
  const struct timeval &timeStamp) throw() {
  //---------------------------------------------------------------------------
  // Note that we do here part of the work of the real syslog call, by building
  // the message ourselves. We then only call a reduced version of syslog
  // (namely castor_syslog). The reason behind it is to be able to set the
  // message timestamp ourselves, in case we log messages asynchronously, as
  // we do when retrieving logs from the DB
  //----------------------------------------------------------------------------

  char   buffer[LOG_MAX_LINELEN * 2];
  size_t len = 0;

  // Do nothing if the logging interface has not been initialized
  if (!s_initialized) {
    return;
  }

  // Do nothing if the log priority is not valid
  if (priority > LOG_DEBUG) {
    return;
  }

  // Initialize buffers
  memset(buffer,  '\0', sizeof(buffer));

  // Start message with priority, time, program and PID (syslog standard
  // format)
  len += build_syslog_header(buffer, s_maxmsglen - len, priority | LOG_LOCAL3,
    timeStamp, getpid());

  // Append the log level, the thread id and the message text
#ifdef __APPLE__
  len += snprintf(buffer + len, s_maxmsglen - len, "LVL=%s TID=%d MSG=\"%s\" ",
    s_prioritylist[priority].text,(int)mach_thread_self(), msg.c_str());
#else
  len += snprintf(buffer + len, s_maxmsglen - len, "LVL=%s TID=%d MSG=\"%s\" ",
    s_prioritylist[priority].text,(int)syscall(__NR_gettid), msg.c_str());
#endif

  // Process parameters
  for (int i = 0; i < numParams; i++) {
    const Param &param = params[i];

    // Check the parameter name, if it's an empty string set the value to
    // "Undefined".
    const std::string name = param.getName() == "" ? "Undefined" :
      _clean_string(param.getName(), true);

    // Process the data type associated with the parameter
    switch (params[i].getType()) {
    // Strings
    case LOG_MSG_PARAM_TPVID:
    case LOG_MSG_PARAM_STR:
      {
        const std::string value = _clean_string(param.getStrValue(), false);
        if (LOG_MSG_PARAM_TPVID == param.getType()) {
          len += snprintf(buffer + len, s_maxmsglen - len, "TPVID=%.*s ",
            CA_MAXVIDLEN, value.c_str());
        } else {
          len += snprintf(buffer + len, s_maxmsglen - len, "%.*s=\"%.*s\" ",
            (int)LOG_MAX_PARAMNAMELEN, name.c_str(),
            (int)LOG_MAX_PARAMSTRLEN, value.c_str());
        }
      }
      break;
    // Numerical values
    case LOG_MSG_PARAM_INT:
      len += snprintf(buffer + len, s_maxmsglen - len, "%.*s=%d ",
        (int)LOG_MAX_PARAMNAMELEN, name.c_str(),
        param.getIntValue());
      break;
    case LOG_MSG_PARAM_INT64:
      len += snprintf(buffer + len, s_maxmsglen - len, "%.*s=%lld ",
        (int)LOG_MAX_PARAMNAMELEN, name.c_str(),
        param.getUint64Value());
      break;
    case LOG_MSG_PARAM_DOUBLE:
      len += snprintf(buffer + len, s_maxmsglen - len, "%.*s=%f ",
        (int)LOG_MAX_PARAMNAMELEN, name.c_str(),
        param.getDoubleValue());
      break;

    // Subrequest uuid
    case LOG_MSG_PARAM_UUID:
      {
        char uuidstr[CUUID_STRING_LEN + 1];
        if (Cuuid2string(uuidstr, CUUID_STRING_LEN + 1,
          &param.getUuidValue())) {
          return;
        }

        len += snprintf(buffer + len, s_maxmsglen - len, "SUBREQID=%.*s ",
                      CUUID_STRING_LEN, uuidstr);
      }
      break;

    case LOG_MSG_PARAM_RAW:
      len += snprintf(buffer + len, s_maxmsglen - len, "%s ",
        param.getStrValue().c_str());
      break;

    default:
      // Please note that this case is used for normal program execution
      // for the following parameter types:
      //
      //   LOG_MSG_PARAM_UID
      //   LOG_MSG_PARAM_GID
      //   LOG_MSG_PARAM_STYPE
      //   LOG_MSG_PARAM_SNAME
      break; // Nothing
    }

    // Check if there is enough space in the buffer
    if ((int)len >= s_maxmsglen) {
      buffer[s_maxmsglen - 1] = '\n';
      break;
    }
  }

  // Terminate the string
  if ((int)len < s_maxmsglen) {
    len += snprintf(buffer + (len - 1), s_maxmsglen - len, "\n");
  }

  castor_syslog(buffer, len);
}

//-----------------------------------------------------------------------------
// writeMsg
//-----------------------------------------------------------------------------
void castor::log::writeMsg(
  const int severity,
  const std::string &msg,
  const int numParams,
  const castor::log::Param params[]) throw() {

  struct timeval timeStamp;
  gettimeofday(&timeStamp, NULL);

  writeMsg(severity, msg, numParams, params, timeStamp);
}
